bitbots_dynup/src/dynup_node.cpp

```diff
@@ -1,112 +1,191 @@
+  ik_.reset();
+  stabilizer_.reset();
+  last_ros_update_time_ = 0;
+  if (std::optional<std::tuple<geometry_msgs::Pose, geometry_msgs::Pose, geometry_msgs::Pose, geometry_msgs::Pose>> poses = getCurrentPoses()) {
     DynupRequest request;
-    request.l_foot_pose = poses->first;
-    request.front = true;
-    request.trunk_pose = poses->second;
-    engine_.setGoals(request); //todo we are currently only getting up from squad
-    stabilizer_.reset();
-    loopEngine();
+    request.l_foot_pose = std::get<0>(poses.value());
+    request.direction = goal->direction;
+    request.r_foot_pose = std::get<1>(poses.value());
+    request.l_hand_pose = std::get<2>(poses.value());
+    request.r_hand_pose = std::get<3>(poses.value());
+    engine_.setGoals(request);
+    if(debug_) {
+      visualizer_.displaySplines(engine_.getRFootSplines(), "base_link");
+      visualizer_.displaySplines(engine_.getLFootSplines(), "r_sole");
+      // Workaround for an error in the Visualizer. TODO
+      if(request.direction == "front" || request.direction == "back") {
+          visualizer_.displaySplines(engine_.getLHandSplines(), "base_link");
+          visualizer_.displaySplines(engine_.getRHandSplines(), "base_link");
+      }
+    }
+    ros::Rate loop_rate(engine_rate_);
+    loopEngine(loop_rate);
     bitbots_msgs::DynUpResult r;
     r.successful = true;
     server_.setSucceeded(r);
   } else {
-    ROS_ERROR_STREAM("Could not determine foot positions! Aborting standup.");
+    ROS_ERROR("Could not determine positions! Aborting standup.");
     bitbots_msgs::DynUpResult r;
     r.successful = false;
     server_.setAborted(r);
   }
 }
 
-void DynUpNode::loopEngine() {
+double DynUpNode::getTimeDelta() {
+  // compute actual time delta that happened
+  double dt;
+  double current_ros_time = ros::Time::now().toSec();
+
+  // first call needs to be handled specially
+  if (last_ros_update_time_==0){
+    last_ros_update_time_ = current_ros_time;
+    return 0.001;
+  }
+  dt = current_ros_time - last_ros_update_time_;
+  // this can happen due to floating point precision
+  if (dt == 0) {
+    ROS_WARN("dt was 0. this can happen in simulation if your update rate is higher than the simulators.");
+    dt = 0.001;
+  }
+  last_ros_update_time_ = current_ros_time;
+  return dt;
+}
+
+void DynUpNode::loopEngine(ros::Rate loop_rate) {
+  int failed_tick_counter = 0;
+  double dt;
   /* Do the loop as long as nothing cancels it */
   while (server_.isActive() && !server_.isPreemptRequested()) {
-    DynupResponse response = engine_.update(1.0 / engine_rate_);
-    DynupResponse stabilized_response = stabilizer_.stabilize(response, ros::Duration(1.0 / engine_rate_));
+    dt = getTimeDelta();
+    DynupResponse response = engine_.update(dt);
+    stabilizer_.setTransforms(tf_buffer_.lookupTransform("r_sole", "base_link", ros::Time(0))); //todo this should not be done based on the tf buffer but on the current open loop goal state
```

Comment:
I partially agree with the comment, this might be rather slow. But I don't know if using the open loop goal state is sufficient, because the stabilizing is performed based on this transform. Maybe just open an issue for that.